datafusion/physical-expr/src/aggregate/first_last.rs

```diff
@@ -165,6 +165,8 @@ struct FirstValueAccumulator {
     orderings: Vec<ScalarValue>,
     // Stores the applicable ordering requirement.
     ordering_req: LexOrdering,
+    // Whether merge_batch() is called before
```

Comment:
Removing those fails `run_first_last_multi_partitions` with `Error: ArrowError(InvalidArgumentError("number of columns(5) must match number of fields(4) in schema"))` (at least in my local test). 

The existing code is somewhat stateful, meaning calling `state()` keeps adding a new column every time to the schema. With this PR, the final aggregation calls `merge_batch()` then, if applicable, calls `state()` for spilling via `GroupedHashAggregateStream.emit()`. The existing code is assuming that `state()` is called only once for partial aggregation. Ideally, calling `state()` multiple times should be allowed without worrying about the internal state.

Reply:
Hmm, why `state` call keeps adding new columns? I think it creates a new `result` and pushes values in and returns it. `result` is not share between `state` calls.

Reply:
`| a | b |` => partial aggregate => `| a | b | is_set |`  => final aggregate (spill calls state()) => `| a | b | is_set |`  => re-grouping => Done

This change avoids doubly adding `| is_set |` for spilling

Reply:
When `spill` is called, `emit` will be called on accumulators which calls `state` as you said. But `state`  returns a fixed vector of `ScalarValue` (e.g., for `FirstValueAccumulator` it is `[first, orderings, is_set]`). I don't see where it adds double `is_set`.

Reply:
I still have question about this. So marked it as unresolved.

Reply:
The strange part is, you only skip `is_set` in state. But actually `state` adds `first`, `orderings` too.
If your theory is `state` called twice to add doubly, `first` and `orderings` should be added doubly too.

And for other accumulators, it is same. Other accumulators might have more than one field in its state.
The strange thing is only `is_set` is skipped. And I don't see why `is_set` is so special.

Reply:
`orderings` is a vector. In reality it will be `[fisrt, orderings[0], orderings[1]..., is_set]`.
Now when final aggregation spill calls this state, `is_set` is already a part of the `orderings`

Reply:
I figured out the reason. This is actually a hack way to fix it. I proposed https://github.com/apache/arrow-datafusion/pull/7565 for this issue.